text/0000-network-update-loop.md

```diff
@@ -0,0 +1,712 @@
+
+# Summary
+[summary]: #summary
+
+Often there is a need to update netcode systems like RPC queue, transport, and others outside the standard [`MonoBehaviour` event cycle](https://docs.unity3d.com/Manual/ExecutionOrder.html).
+
+This RFC proposes a new **Network Update Loop** infrastructure that utilizes [Unity's low-level Player Loop API](https://docs.unity3d.com/ScriptReference/LowLevel.PlayerLoop.html) and allows for registering `INetworkUpdateSystem`s with `NetworkUpdate()` methods to be executed at specific `NetworkUpdateStage`s which may also be before or after `MonoBehaviour`-driven game logic execution.
+
+Implementation is expected to have a minimal yet flexible API that would allow further systems such as network tick to be easily developed.
+
+# Motivation
+[motivation]: #motivation
+
+Even though it is possible to use low-level Player Loop API directly to insert custom `PlayerLoopSystem`s into the existing player loop, it also requires a non-trivial amount of boilerplate code and could cause fundamental issues in the engine runtime if not done very carefully. **We would like to have less boilerplate code and also be less error-prone at runtime.**
+
+Beyond that, the proposed design standardizes `NetworkUpdateStage`s which are going to be executed at specific points in the player loop. This allows other framework systems tied into `NetworkUpdateLoop` to be aligned, such as RPCs executing at specific stages. **We would like to standardize the execution of network stages as an infrastructure in the framework.**
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Developers will mostly interact with `NetworkUpdateLoop` for registration and `INetworkUpdateSystem` for implementation.
+
+## NetworkUpdateLoop
+
+`NetworkUpdateLoop` implements custom methods injected into the player loop, maintains stage-specific lists of `INetworkUpdateSystem`s.
+
+### Registration
+
+`NetworkUpdateLoop` exposes 4 methods for registration:
+
+- `void RegisterNetworkUpdate(INetworkUpdateSystem updateSystem, NetworkUpdateStage updateStage)`
+  - registers an `INetworkUpdateSystem` to be executed on the specified `NetworkUpdateStage`
+- `void RegisterAllNetworkUpdates(INetworkUpdateSystem updateSystem)`
+  - registers an `INetworkUpdateSystem` to be executed on all `NetworkUpdateStage`s
+- `void UnregisterNetworkUpdate(INetworkUpdateSystem updateSystem, NetworkUpdateStage updateStage)`
+  - unregisters an `INetworkUpdateSystem` from the specified `NetworkUpdateStage`
+- `void UnregisterAllNetworkUpdates(INetworkUpdateSystem updateSystem)`
+  - unregisters an `INetworkUpdateSystem` from all `NetworkUpdateStage`s
+
+### Frames and Stages
+
+After injection, player loop will look like this:
+
+- **Initialization**
+  - // other systems
+  - RunNetworkInitialization
+    - `NetworkLoopSystem.AdvanceFrame()`
```

Comment:
Feels like frame id would be an INetworkUpdateSystem itself, not a core part of the base loop? Why would we want to match the network frame count to the runtime frame count and not to, e.g., when networking is initialized? Maybe I'm reading too much into the psuedo-code here?

Reply:
no, I think you are right. frame concept should be a system on its own and not a part of the network update loop.